contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,264 @@
+        address observer = tokenObservers[tokenId];
+        if (observer != address(0)) {
+            PermissionedRegistryTokenObserver(observer).onRelinquish(tokenId, msg.sender);
+        }
+
+        emit NameRelinquished(tokenId, msg.sender);
+    }
+
+    function setUri(uint256 tokenId, string memory _uri) 
         external
         onlyTokenOwner(tokenId)
-        withSubregistryFlags(tokenId, FLAG_SUBREGISTRY_LOCKED, 0)
+    {
+        metadataProvider.setTokenUri(tokenId, _uri);
+        emit URI(_uri, tokenId); 
+    }
+
+    function getSubregistry(string calldata label) external view virtual override(BaseRegistry, IRegistry) returns (IRegistry) {
+        (address subregistry, uint96 flags) = datastore.getSubregistry(uint256(keccak256(bytes(label))));
+        uint64 expires = _extractExpiry(flags);
+        if (expires <= block.timestamp) {
+            return IRegistry(address(0));
+        }
+        return IRegistry(subregistry);
+    }
+
+    function getResolver(string calldata label) external view virtual override(BaseRegistry, IRegistry) returns (address) {
+        uint256 tokenId = uint256(keccak256(bytes(label)));
+        (, uint96 flags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(flags);
+        if (expires <= block.timestamp) {
+            return address(0);
+        }
+
+        (address resolver, ) = datastore.getResolver(tokenId);
+        return resolver;
+    }
+
+    function setSubregistry(uint256 tokenId, IRegistry registry)
+        external
+        onlyRoles(tokenIdResource(tokenId), ROLE_SET_SUBREGISTRY)
     {
         (, uint96 _flags) = datastore.getSubregistry(tokenId);
         datastore.setSubregistry(tokenId, address(registry), _flags);
     }
 
     function setResolver(uint256 tokenId, address resolver)
         external
-        onlyTokenOwner(tokenId)
-        withSubregistryFlags(tokenId, FLAG_RESOLVER_LOCKED, 0)
+        onlyRoles(tokenIdResource(tokenId), ROLE_SET_RESOLVER)
     {
         (, uint96 _flags) = datastore.getResolver(tokenId);
         datastore.setResolver(tokenId, resolver, _flags);
     }
 
-    function flags(uint256 tokenId) external view returns(uint96) {
+    function setFlags(uint256 tokenId, uint96 _flags)
+        external
+        onlyRoles(tokenIdResource(tokenId), ROLE_SET_FLAGS)
+        returns (uint256 newTokenId)
+    {
+        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint96 newFlags = oldFlags | (_flags & FLAGS_MASK);
+
+        if (newFlags != oldFlags) {
+            datastore.setSubregistry(tokenId, subregistry, newFlags);
+
+            newTokenId = (tokenId & ~uint256(FLAGS_MASK)) | (newFlags & FLAGS_MASK);
+            if (tokenId != newTokenId) {
+                address owner = ownerOf(tokenId);
+                _mint(owner, newTokenId, 1, "");
+                _burn(owner, tokenId, 1);
+            }
+        } else {
+            newTokenId = tokenId;
+        }
+    }
```

Comment:
Are we 100% sure we still need flags, now that we have the new permission model?

Reply:
Well, I wasn't sure what else you guys wanted flags for so left them in...should I remove them? This would also mean being able to cleanup the "token id changes when flags change" logic, right?

Reply:
If we're not using them now, then yes we can.

The issue the token ID change was created to fix is present here in another form - we don't want a situation where someone can grant themselves permissions just prior to transferring the domain to the new owner, and we need a way to make sure the user has 'clean title' to the name they just bought. Needs some thought on how to handle this.

Reply:
See https://github.com/ensdomains/namechain/pull/29#issuecomment-2760497409